Add draft groups

alexaunter opened the pull request:
Implement draft groups
new_draft_group
get_draft_groups_paged
new_draft
get_draft
get_drafts
add new message type for ft_token_receiver
add convert_draft method
add new_drafts (batch method)
add convert_drafts (batch method)

alexaunter committed:
feat(drafts): add drafts and groups create and view methods

alexaunter committed:
feat(drafts): add draft group funding

alexaunter committed:
feat(drafts): add draft converting

alexaunter committed:
feat(drafts): add draft method batching

frol commented on src/lib.rs:
```diff
@@ -196,4 +210,68 @@ impl Contract {
+        index
+    }
+
+    pub fn new_drafts(&mut self, drafts: Vec<Draft>) -> Vec<DraftIndex> {
```
I would use a verb to indicate that this is an action: `create_drafts` (and similarly `create_draft`)

frol commented on src/draft.rs:
```diff
@@ -0,0 +1,51 @@
+#[serde(crate = "near_sdk::serde")]
+pub struct Draft {
+    pub draft_group_id: DraftGroupIndex,
+    pub lockup_id: Option<LockupIndex>,
+    pub lockup: Lockup,
+}
+
+impl Draft {
+    pub fn assert_new_valid(&self) {
+        assert_eq!(self.lockup_id, None, "new draft must not have lockup_id");
+
+        let amount = self.lockup.schedule.total_balance();
+        self.lockup.assert_new_valid(amount);
+    }
+
+    pub fn assert_can_convert(&self) {
+        assert!(self.lockup_id.is_none(), "draft already converted");
+    }
+}
```
1. I suggest using Rust enums to avoid impossible state to be reachable and avoid having two copies of the Lockup in the state:

```rust
enum DraftOrConvertedLockup {
    Draft(Lockup),
    Converted(LockupIndex),
}
```

This way you won't need to check on lockup_id, and you won't be able to forget to do it.

2. Why do we even need to store the converted lockups? It seems that we never remove the drafts and draft groups. I don't think it makes sense to keep drafts in the storage forever. In order to list history of lockups we planned to use Indexer (which will also be able to attach timestamp, etc)

alexaunter committed:
refactor(drafts): replace vector with unordered map

alexaunter committed:
feat(drafts): delete drafts after convert, remove lockup_id tracking

alexaunter committed:
refactor(ft_token_receiver): use untagged enum for incoming message

alexaunter committed:
feat(convert): remove index and total amount on convert

alexaunter committed:
refactor(convert): add hepler method total_balance to drafts

alexaunter committed:
refactor(termination): rename terminator_id to payer_id

alexaunter committed:
feat(termination): always return money to lockup/draft creator

alexaunter committed:
feat(draft): update payer_id before lockup conversion

frol commented on src/ft_token_receiver.rs:
```diff
@@ -13,16 +27,38 @@ impl FungibleTokenReceiver for Contract {
+        let ft_message: FtMessage = serde_json::from_str(&msg).unwrap();
+        match ft_message {
+            FtMessage::Lockup(lockup) => {
+                // TODO: rework
```
What is this TODO about?

frol commented on src/termination.rs:
```diff
@@ -13,26 +13,21 @@ pub enum HashOrSchedule {
 pub struct TerminationConfig {
     /// The account ID that can terminate vesting.
-    pub terminator_id: ValidAccountId,
+    pub payer_id: ValidAccountId,
```
Let's call it `beneficiary_id` (similarly how native DELETE_ACCOUNT action on NEAR Protocol defines this terminology)

frol commented on src/lib.rs:
```diff
@@ -199,4 +212,80 @@ impl Contract {
+        if let Some(mut termination_config) = draft.lockup.termination_config {
+            termination_config.payer_id = draft_group.payer_id.unwrap();
+            draft.lockup.termination_config = Some(termination_config);
+        }
```
Do I miss something or you could just borrow mutably here?

```suggestion
        if let Some(&mut termination_config) = draft.lockup.termination_config {
            termination_config.payer_id = draft_group.payer_id.unwrap();
        }
```

frol reviewed (commented):
I only have a few nits in the comments, and one feature-request:

Let's remove the draft group once it is completely converted (ensure that the draft group ids are not changed when one group is removed, so approval requests always point to the id that is immutable like you did for the draft ids already).

frol commented:
@evgenykuzyakov The context for this PR is that we want to be able to approve one transaction through DAO, and then progress on converting the drafted lockups to the list of funded lockups from any account without a need to involve DAO/multisig in this process.

alexaunter committed:
refactor: drop support for optional vesting schedule

alexaunter committed:
refactor(lockups): add LockupCreate model for drafts

alexaunter committed:
refactor: cargo fmt

alexaunter committed:
refactor: add lockup create view

alexaunter committed:
feat: use LockupCreate to create individual lockups

alexaunter committed:
test(terminate): bring back same as lockup scchedule vesting option

alexaunter committed:
refactor: rename payer_id to beneficiary_id

alexaunter committed:
feat(drafts): remove draft group after all drafts are converted

alexaunter committed:
feat(view): expose next_draft_group_id

alexaunter committed:
refactor: add fund draft_group helper method

alexaunter committed:
refactor(drafts): remove draft.assert_new_valid useless parameter

alexaunter commented:
@frol , I've fixed previous issues and implemented cleaning of draft_group when it's fully converted.
Also I decided to created a separate model LockupCreate, because storing Lockup inside a draft with possibly incorrect lockup id and patching it when creating the actual lockup feels hacky. Waiting for the feedback.

frol commented on src/termination.rs:
```diff
@@ -13,31 +14,23 @@ pub enum HashOrSchedule {
 #[cfg_attr(not(target_arch = "wasm32"), derive(Debug, PartialEq, Clone))]
 pub struct TerminationConfig {
     /// The account ID that can terminate vesting.
```
Let's update this comment

frol reviewed (approved):
> Also I decided to created a separate model LockupCreate, because storing Lockup inside a draft with possibly incorrect lockup id and patching it when creating the actual lockup feels hacky.

This is a good clean solution. Thanks for thinking it through!

This PR looks good to me now

alexaunter committed:
docs(termination): update field comment

evgenykuzyakov commented on src/lib.rs:
```diff
@@ -145,12 +160,12 @@ impl Contract {
         hashed_schedule: Option<Schedule>,
     ) -> PromiseOrValue<WrappedBalance> {
-        let account_id = env::predecessor_account_id();
+        self.assert_deposit_whitelist(&env::predecessor_account_id());
```
While I see the intention, I don't think it makes sense to mix the two groups together. I'd prefer if the termination can only be called by either the account that initialized (or funded) the lockup, or the specified account.
The whitelist for deposits is different from terminations.

evgenykuzyakov commented on src/draft.rs:
```diff
@@ -0,0 +1,65 @@
+pub struct DraftGroup {
+    pub total_amount: Balance,
+    pub beneficiary_id: Option<ValidAccountId>,
+    pub draft_indices: HashSet<DraftIndex>,
```
Is there a case where the number of drafts exceeds gas? Should we use another `UnorderedSet` here?

alexaunter replied:
According to my benchmarks
read + update + save hashset of 10000 values eats 23 TGas
https://explorer.testnet.near.org/transactions/CU4jiQWtX2kWx6rWvneFF1WwgqtEro3mR5hwJGvfoys4

read + update + save hashset of 50000 values eats 105 TGas
https://explorer.testnet.near.org/transactions/5gtSos2icwpuuZWAdJQaGRGu5KAaT7JCtYV3ABdXroHE

What orders of magnitude should we target?
FYI @frol

evgenykuzyakov replied:
Should be good.

alexaunter replied:
I've fixed gas usage for hashset repacking in batch draft methods (create_drafts, convert_drafts)

evgenykuzyakov commented on src/lib.rs:
```diff
@@ -199,4 +214,90 @@ impl Contract {
+            .into_iter()
+            .map(|draft_id| self.convert_draft(draft_id))
+            .collect()
+    }
```
Using `HashSet` in the `DraftGroup`, the deserialization/serialization will be quite expensive for multiple drafts. Switching to `UnorderedSet` in the `DraftGroup` or have benefits. Another option is cache it somehow, or refactor 2 methods by extracting deserialization/serialization out.

evgenykuzyakov commented on src/termination.rs:
```diff
@@ -1,9 +1,10 @@
 #[serde(crate = "near_sdk::serde")]
-#[cfg_attr(not(target_arch = "wasm32"), derive(Debug, PartialEq, Clone))]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Debug, PartialEq))]
 pub enum HashOrSchedule {
```
Maybe rename this.

evgenykuzyakov commented on src/termination.rs:
```diff
@@ -1,9 +1,10 @@
-#[cfg_attr(not(target_arch = "wasm32"), derive(Debug, PartialEq, Clone))]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Debug, PartialEq))]
 pub enum HashOrSchedule {
+    SameAsLockupSchedule,
```
This is a breaking state change, which makes lockup non-upgradable. So if there are some lockups already deployed it can't be migrated to this version.

alexaunter replied:
@evgenykuzyakov, honestly, I see no point keeping the backward compatibility right now.
There is no self upgrade functionality inside the current contract, so the only option is to upgrade it via full access key.
In that case, any code can be deployed, any transaction can be signed on behalf of the lockup contract. It means the lockup contract does not serve it's purpose (it does not guarantee that locked tokens will be available for you to claim).

@frol , does anyone use this lockup code in production?
Do we have any plans to allow the upgrades?
Do we need to keep backward compatibility at this stage?

I thought that instead of upgrades we would rather create a new lockup contract with new code, totally unrelated to the original one.

frol replied:
As we discussed with @evgenykuzyakov, there is one known production use of this contract, but I agree that without proper upgradability we can ignore the breaking API concern. @evgenykuzyakov suggested offline that we bump the version to indicate the breaking change and call it a day, and I agree with him.

alexaunter replied:
Bumped the version.

evgenykuzyakov commented on src/view.rs:
```diff
@@ -109,4 +190,35 @@ impl Contract {
+        let from_index = from_index.unwrap_or(0);
+        let to_index = to_index.unwrap_or(self.draft_groups.len() as _);
+        (from_index..std::cmp::min(self.next_draft_group_id as _, to_index))
+            .filter_map(|index| self.get_draft_group(index).map(|group| (index, group)))
```
Better to iterate on the vector of keys/values. You can use `keys_as_vector` and `values_as_vector` to iterate only on existing groups. For this you may also need to expose actual .len() as a view method.

alexaunter replied:
Updated. @evgenykuzyakov , do you think it's still valid to call func args from_index, to_index? Should they be renamed?

alexaunter committed:
Merge remote-tracking branch 'upstream/main' into add_draft_groups_2

evgenykuzyakov reviewed (commented):
Sorry, forgot to click submit review.

alexaunter committed:
feat(termination): revert previous behavior: only payer can terminate

alexaunter committed:
refactor: rename HashOrSchedule to VestingConditions

alexaunter committed:
chore: bump contract version

alexaunter committed:
tests(terminate): improve convert batch test

alexaunter committed:
refactor(configs): reduce gas in convert batch

alexaunter committed:
feat(view): rework draft indices paging

alexaunter committed:
Merge remote-tracking branch 'upstream/main' into add_draft_groups_2

alexaunter committed:
perf(drafts): reduce gas usage in create_drafts batch method

alexaunter committed:
style: cargo fmt batch drafts methods

alexaunter committed:
refactor(drafts): rewrite create drafts with memoizing draft groups

alexaunter committed:
refactor(drafts): rewrite convert drafts with memoizing draft groups

alexaunter committed:
style: cargo fmt

alexaunter committed:
Merge remote-tracking branch 'upstream/main' into add_draft_groups_2

alexaunter committed:
refactor: rename assert_can_convert for consistency

alexaunter committed:
refactor: rename beneficiary_id to payer_id for cleaner terminology

alexaunter committed:
fix: add explicit overflow check to draft create

alexaunter committed:
refactor: remove LockupCreate new_unlocked method from wasm binary

alexaunter committed:
refactor: move existing fields to the top

alexaunter committed:
refactor: add comment on get_draft_groups_paged params

alexaunter committed:
style: remove newline

alexaunter committed:
refactor: remove serde(default) for view

alexaunter committed:
refactor: remove unused clone

alexaunter committed:
refactor: fix typo

alexaunter committed:
refactor: remove unnecessary derive

alexaunter committed:
style: cargo fmt

alexaunter committed:
refactor: remove unnecessary serde default

alexaunter committed:
refactor: rename view field to match the original model

alexaunter committed:
refactor: rearrange line s for readability

alexaunter committed:
refactor: use beneficiary_id field name in termination config

evgenykuzyakov commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ft-lockup"
-version = "0.2.0"
+version = "0.3.0"
```
Since it's a breaking change.
```suggestion
version = "1.0.0"
```

alexaunter replied:
@evgenykuzyakov , are we following the semver here? 
quoting https://semver.org/, anything can be changed before 1.0.0 release, but I'm okay with moving forward.
```
Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.
```

alexaunter committed:
chore: bump version to 1.0.0

Co-authored-by: Evgeny Kuzyakov <h3r0k1ll3r@gmail.com>

evgenykuzyakov commented:
Please rebuild with docker

alexaunter committed:
chore: rebuild the binary

alexaunter commented:
@evgenykuzyakov , wasm file has been rebuilt
